File: msad2023_pri/Plotter.cpp
/*
    Plotter.cpp

    Copyright © 2021 Wataru Taniguchi. All rights reserved.
*/
#include "Plotter.hpp"

Plotter::Plotter(ev3api::Motor* lm, ev3api::Motor* rm, ev3api::GyroSensor* gs) :
leftMotor(lm),rightMotor(rm),gyroSensor(gs) {
    /* reset motor encoders */
    leftMotor->reset();
    rightMotor->reset();
    /* reset gyro sensor */
    gyroSensor->reset();
    /* initialize variables */
    prevAngL = leftMotor->getCount();
    prevAngR = rightMotor->getCount();
    distance = azimuth = locX = locY = 0.0;
}

int Plotter::getDistance() {
    return (int)distance;
}

int Plotter::getAzimuth() {
    return (int)azimuth;
}

int Plotter::getDegree() {
    // degree = 360.0 * radian / M_TWOPI;
    double degree = (360.0 * azimuth / M_TWOPI);
    if (degree == 360) degree = 0; /* guarantee 0 <= degree <= 359 */ 
    return (int)degree;
}

int Plotter::getLocX() {
    return (int)locX;
}

int Plotter::getLocY() {
    return (int)locY;
}

int Plotter::getAngL() {
    return prevAngL;
}

int Plotter::getAngR() {
    return prevAngR;
}

void Plotter::setDegree(int deg) {
  // double radian = degree * M_TWOPI / 360.0
  azimuth = (double) deg * M_TWOPI / 360.0;
}

void Plotter::plot() {
    /* accumulate distance */
    int32_t curAngL = leftMotor->getCount();
    int32_t curAngR = rightMotor->getCount();
    double deltaDistL = M_PI * TIRE_DIAMETER * (curAngL - prevAngL) / 360.0;
    double deltaDistR = M_PI * TIRE_DIAMETER * (curAngR - prevAngR) / 360.0;
    double deltaDist = (deltaDistL + deltaDistR) / 2.0;
    if (deltaDistL * deltaDistR >= 0) { /* cumulate only when both wheels spin in the same direction */  
      if (deltaDist >= 0) { /* cumulative distance must be always positive */
	distance += deltaDist;
      } else {
	distance -= deltaDist;
      }
    }
    prevAngL = curAngL;
    prevAngR = curAngR;
    /* calculate azimuth */
    double deltaAzi = (deltaDistL - deltaDistR) / WHEEL_TREAD;
    azimuth += deltaAzi;
    if (azimuth >= M_TWOPI) {
        azimuth -= M_TWOPI;
    } else if (azimuth < 0.0) {
        azimuth += M_TWOPI;
    }
    if (deltaDistL * deltaDistR >= 0) { /* change location only when both wheels spin in the same direction */  
      /* estimate location */
      locX += (deltaDist * sin(azimuth));
      locY += (deltaDist * cos(azimuth));
    }
}

File: 2024base/py_etrobo_util/plotter.py
import math
from etrobo_python import ETRobo, Hub, Motor, ColorSensor, TouchSensor, SonarSensor, GyroSensor

TIRE_DIAMETER: float = 100.0
WHEEL_TREAD: float = 120.0

class Plotter(object):
    def __init__(self) -> None:
        self.running = False
        self.distance = 0.0

    def plot(
        self,
        hub: Hub,
        arm_motor: Motor,
        right_motor: Motor,
        left_motor: Motor,
        touch_sensor: TouchSensor,
        color_sensor: ColorSensor,
        sonar_sensor: SonarSensor,
        gyro_sensor: GyroSensor,
    ) -> None:
        if not self.running:
            self.running = True
            right_motor.reset_count()
            left_motor.reset_count()
            self.prev_ang_r = right_motor.get_count()
            self.prev_ang_l = left_motor.get_count()
            return

        cur_ang_r = right_motor.get_count()
        cur_ang_l = left_motor.get_count()
        delta_dist_r = math.pi * TIRE_DIAMETER * (cur_ang_r - self.prev_ang_r) / 360.0
        delta_dist_l = math.pi * TIRE_DIAMETER * (cur_ang_l - self.prev_ang_l) / 360.0
        delta_dist = (delta_dist_r + delta_dist_l / 2.0)
        if (delta_dist >= 0.0):
            self.distance += delta_dist
        else:
            self.distance -= delta_dist
        self.prev_ang_r = cur_ang_r
        self.prev_ang_l = cur_ang_l
        return

    def get_distance(self) -> int:
        return int(self.distance)
